Coco/area relational integrity

CocoisBuggy opened the pull request:
Fulfils the feature request in #126 

There's a bit happening here, since I made some performance driven changes to help me with a different PR (#407) I tried to knock out the major test cases but the nature of this particular feature is a little thorny because of how we utilize de-normalized data.

The commit history is a little messy and I am frankly feeling a bit lazy, which is why I have not squashed them. Been moving between machines a bunch and reworked the approach a few times to get speed tuned in.
If someone would like me to squash then I will try and motivate myself to do it.

The key takeaway in terms of changes that will affect other devs is that a new field has been introduced on the area document from which all other ancestry / children are derived. the `parent` field points to the areas direct parent. The pathTokens and ancestors fields have been colocated into a single array of sub documents so that the update propogations can happen all together - not a big deal now, but gives us a sane path to embed  more de-normalized data if we need to.

@vnugent I'd like your particular eyes for these changes.

CocoisBuggy committed:
shebang should be first line

(https://en.wikipedia.org/wiki/Shebang_(Unix)#Version_8_improved_shell_scripts if you are sufficiently nerdy to care)

CocoisBuggy committed:
For those using bash beautify vscode extension

CocoisBuggy committed:
Make replica set specification optional in seed script

I ran into a one-off issue where the replica set specification was
throwing off the script in a development environment. This code will let
you un-set the variable if you don't need it while in dev.

CocoisBuggy committed:
Add a shell.nix for people using nixOS

This is by no means a complete environment, but if no one minds me
leaving this here it would save me some trouble - but is not required.

If you are unfamiliar with nixos, this file is intended to give my
environment the necessary mongo utils / yarn / node and so forth.

You can peg a node version and all sorts of various magic. I would
recommend any nerds with sufficient free time investigate - but you will
find it very hard to escape once it has ensnared you.

CocoisBuggy committed:
Integrate memory server into nixos environment

CocoisBuggy committed:
compass is a neat tool - though duplicates much of the functionality of mono-express

CocoisBuggy committed:
A naieve implementation of the functionality I was thinking of using.

The query is too slow, it remains to be seen if this is because the
scope is simply too liberal - but my intuition is that this is down
to the regex search and full-document retrieval

Current root query time is along the lines of 16 seconds (Unthinkably
slow) so I'll try another approach, otherwise this architechture is
simply no good.

CocoisBuggy committed:
I think I may have been wrong about this. Path Hash is not a cached
value? This may be down to me not manually running some cron operation
in my development environment?

I will work it out later.

CocoisBuggy committed:
Performance improvements and less jank

There is still some query tuning that I want to do - I might sqaush this
commit later

Cost Reference
 Texas ~50ms
 California ~900ms

CocoisBuggy committed:
minor speed improvement to the query

achieved by pruning the data being passed through the pipeline

CocoisBuggy committed:
minor speed improvement to the query

achieved by pruning the data being passed through the pipeline

CocoisBuggy committed:
Tests added and param options implemented

The only way I can think of to speed this up is to eliminate the parent
lookup stage in the pipeline, which seems to double the time the query
takes to run until completion.

The natural way to address this would be to embed a parent pointer -
which I assume should be easy enough since there must be existing code
to maintain the integrity of the

pathTokens,
ancestors

So I'll take a look at doing that too

CocoisBuggy committed:
Merge branch 'coco-hierarchy-tools' of github.com:OpenBeta/openbeta-graphql into coco-hierarchy-tools

CocoisBuggy committed:
Add utility for using or creating a client session

Rather than duping logic from other code units over and over

CocoisBuggy committed:
Merge branch 'coco/#409' into coco/area-relational-integrity

CocoisBuggy committed:
update script to allow 'mongosh'

the name of mongo was changed to mongo a little while ago, and I can't
be bothered to symlink the name, so now if mongo is not available the
script will fallback onto mongosh.

+ Add mongosh to the nix shell

CocoisBuggy committed:
Migrate to new data model

The tests are not passing at this commit, even though TSC is not
reporting type issues. I will try and correct whatever is preventing the
types from being reported.

CocoisBuggy committed:
Allow compass to open on the correct connection

Previously you would need to re-enter the auth credentials.

CocoisBuggy committed:
Begin integreating the parent-centric changes

This commit should be squashed later...

CocoisBuggy committed:
Add debug flag to tests

CocoisBuggy committed:
Add an attach request to the launch options

CocoisBuggy committed:
Prefer optional dbg over forced

CocoisBuggy committed:
I believe we have moved passed the need for import

This code, as far as I can tell, hasn't been touched for a while and is
just clutter? If I am wrong about this then I need to know why we are
still importing data that we should presumably now be maintaining.

CocoisBuggy committed:
Denormalize related data together

I'm not sure why I didn't do this initially

CocoisBuggy committed:
Merge branch 'develop' into coco/area-relational-integrity

CocoisBuggy committed:
Area ancestry deconvolution

There is a new, simple, source of truth for area structure derivations.
Each area document holds a parent reference, which points to its desired
reference (or none, if it is a root node). The derived fields such as

 - Children
 - Ancestry

have been collated in `embeddedRelations`. the denormalized fields
`ancestors` and `pathTokens` have also been colocated into a single
entity path that makes propogating updates to denormalized data easier
(e.g, when updating an areas name). It leaves us a space for adding new
fields, and finally exposes the ACTUAL id of the document so that we can
index effectively.

I would suggest that using the parent field is the most straightforward
and safest thing to do in pretty much all scenarios, but if you are a
mongodb veteran you may find need for the array indexing features -
though as far as I can tell, this is rarely the case?

 - Added tests
 - Updated some existing tests
 - Data returned at the gql layer should be identical, but I have not
   fully de-risked those tests. All I can say is that they are passing.

This commit does not address #126, but gives us an enviable headstart
at it. It is however needed by #407

CocoisBuggy committed:
Checkpoint commit -- pls squash

I'm going to go live in the mountains for a week or so, but will take a
laptop so may want to work on this a bit.

CocoisBuggy committed:
Checkpoint - back from mountains

CocoisBuggy committed:
Area parent migration implemented and tested

CocoisBuggy committed:
Expose functionality via the gql layer

It's all in, I think. I added a new test to cover name uniqueness but no
doubt more cases will come up.

CocoisBuggy committed:
Merge with dependency updates

CocoisBuggy committed:
Merge branch 'develop' into coco/area-relational-integrity

CocoisBuggy committed:
Fix double import (oops)

vnugent commented:
@CocoisBuggy can you please explain what problems this PR is trying to solve?  Maybe a diagram or example would help.  Thanks!
Not sure if you notice I created a rough plan for v2 #422 where we can address some of the major issues.  Perhaps we can discuss larger impact changes in v2?

CocoisBuggy commented:
Sure thing, sorry I suppose I wasn't super clear in my explanation.


The tldr is:
- performance
- code sanity
- collate demoralized fields together
- allow updating of area parents to reorganize areas
- deconvolute data structure

## Who should be affected?

Only developers working on the back end. There is a scary number of changes, but I haven't touched the way in which the data is accessed by the api so there **shouldn't** be a problem - but now that I'm writing that it may be just as well for me to check that the existing tests cover the behavior.

Developers shouldn't have to change their mental model too much, since all the old data Is still available - and is recomputed when it's dependencies update.

## Performance

In my other pr I'm working on some queries that I need for a feature I'm interested in, but performing the lookup for parents of nodes was always a substantial hit when I was looking at my profiler. Partly this is a consequence of storing the ancestors as a comma delimited string which seemed to always be a performance sink whenever I touched it. When using the children array I found a couple of strange hydras when performing deep queries.

After struggling to get the queries down below 100-200ms I decided to look elsewhere for a solution.

## Code Sanity

While I was looking into area update and coming to terms with the existing denormalization strategies I found myself yearning for a single approach to be used so that it could all be debugged and understood together - especially since they're absolutely related code work.

It also made me increasingly perplexed that the children were specified as an array of ObjectId but the ancestors is a string - and a string of uuid values at that.

My solution to future-proof against this problem getting worse over time and also resolve the inconsistencies I decided to put the related data together.

## Area Structure Updating

When I went to implement this feature neither the array of children nor the ancestor string felt like good ways to express which area had which parent. I decided to make these explicitly derived fields, and then simply put a parent field on each area. The difference here is slightly more than philosophical, but without going into too much detail my belief is that this is the easiest way to not get stuck in side effect hell. I can expand more on that if needed, but for the time being it suffices to say that there's no way to accidentally end up with areas belonging to multiple parents, which would have cascading bugs.

So now you need only update an areas parent, and then recompute the paths from the altered node (which is quite speedy).

## conclusion

The pr is more or less worthless if we want to specifically avoid a parent field pattern. It should work well with the v2 milestone, especially for phasing out uuid under the hood. I can do more legwork to make sure it doesn't change the existing api, as it shouldn't be necessary for this pr

vnugent commented:
@CocoisBuggy Thanks for the large PR.  Can we discuss this after the new year?

CocoisBuggy commented:
For sure, merry Christmas @vnugent !

CocoisBuggy committed:
Merge branch 'develop' into coco/area-relational-integrity

CocoisBuggy committed:
Delete Area Transformer

My confidence in this deletion is a little lower than it initially was.
I remember checking the code archeology the first time around and
thinking that we must have been done with importing data (especially
since descriptions have become mutable and I had assumed importing
might result in unintensional overwrites.)

However this change to the code indicates otherwise. Could someone
indicate where in the workflow this code is still in use? If it's still
in use then obviously I actually do need to fix this up for the PR to
make any sense...

Happy new year, fellows!

CocoisBuggy commented:
Happy new year fellow climbers! I hope everyone had a good christmas. Down here in SA we are so far out of season we do a nightly prayer for our poor skin.

I thought I'd resolve the merge issues as a nice little start to the year. I'm going to copy the contents of my previous commit as it contains a question I now have:

```
My confidence in this deletion is a little lower than it initially was.
I remember checking the code archeology the first time around and
thinking that we must have been done with importing data (especially
since descriptions have become mutable and I had assumed importing
might result in unintensional overwrites.)

However this change to the code indicates otherwise. Could someone
indicate where in the workflow this code is still in use? If it's still
in use then obviously I actually do need to fix this up for the PR to
make any sense...
```

CocoisBuggy committed:
Merge branch 'develop' into coco/area-relational-integrity

CocoisBuggy commented:
#316 
#352
#126
#458
are all partially or fully resolved in this pr, any eyes on this or feedback would be greatly appreciated
